lock-api/src/test/java/com/palantir/lock/watch/ClientLockWatchEventLogImplTest.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.palantir.lock.v2.LockToken;
+
+@RunWith(MockitoJUnitRunner.class)
+public final class ClientLockWatchEventLogImplTest {
+    private static final UUID LEADER = UUID.randomUUID();
+    private static final IdentifiedVersion VERSION_1 = IdentifiedVersion.of(LEADER, 17L);
+    private static final IdentifiedVersion VERSION_2 = IdentifiedVersion.of(LEADER, 25L);
+    private static final LockWatchEvent EVENT_1 =
+            LockEvent.builder(ImmutableSet.of(), LockToken.of(UUID.randomUUID())).build(17L);
+    private static final LockWatchEvent EVENT_2 =
+            LockEvent.builder(ImmutableSet.of(), LockToken.of(UUID.randomUUID())).build(25L);
+    private static final LockWatchStateUpdate.Snapshot SNAPSHOT =
+            LockWatchStateUpdate.snapshot(VERSION_1.id(), VERSION_1.version(), ImmutableSet.of(), ImmutableSet.of());
+    private static final LockWatchStateUpdate.Success SUCCESS =
+            LockWatchStateUpdate.success(VERSION_2.id(), VERSION_2.version(), ImmutableList.of(EVENT_1, EVENT_2));
+    private static final LockWatchStateUpdate.Failed FAILED =
+            LockWatchStateUpdate.failed(LEADER);
+
+    @Mock
+    private ClientLockWatchSnapshotUpdater snapshotUpdater;
```

Comment:
Going to read through these later, but given my gut feeling that these 2 classes will be rather coupled, would be good to have a single integration test that tests it end to end.

Reply:
I agree with this, will write an integration test once I'm done with other comments

Reply:
Done